server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -54,135 +57,254 @@
+        private boolean finished;
+
+        Transaction(MetaStateService metaStateService, Manifest previousManifest) {
+            this.metaStateService = metaStateService;
+            assert previousManifest != null;
+            this.previousManifest = previousManifest;
+            this.commitCleanupActions = new ArrayList<>();
+            this.rollbackCleanupActions = new ArrayList<>();
+            this.finished = false;
+        }
+
+        long writeGlobalState(String reason, MetaData metaData) throws WriteStateException {
+            assert finished == false : TRANSACTION_COMPLETED_MSG;
+            try {
+                long generation = metaStateService.writeGlobalState(reason, metaData);
+                commitCleanupActions.add(() -> metaStateService.cleanupGlobalState(generation));
+                rollbackCleanupActions.add(() -> metaStateService.cleanupGlobalState(previousManifest.getGlobalGeneration()));
+                return generation;
+            } catch (WriteStateException e) {
+                rollback();
+                throw e;
             }
+        }
 
+        long writeIndex(String reason, IndexMetaData metaData) throws WriteStateException {
+            assert finished == false : TRANSACTION_COMPLETED_MSG;
+            try {
+                Index index = metaData.getIndex();
+                long generation = metaStateService.writeIndex(reason, metaData);
+                commitCleanupActions.add(() -> metaStateService.cleanupIndex(index, generation));
+                long previousGeneration = previousManifest.getIndexGenerations().getOrDefault(index, -1L);
+                rollbackCleanupActions.add(() -> metaStateService.cleanupIndex(index, previousGeneration));
+                return generation;
+            } catch (WriteStateException e) {
+                rollback();
+                throw e;
+            }
+        }
 
-            relevantIndices = getRelevantIndices(event.state(), event.previousState(), previouslyWrittenIndices);
-            final Iterable<IndexMetaWriteInfo> writeInfo = resolveStatesToBeWritten(previouslyWrittenIndices, relevantIndices,
-                previousMetaData, event.state().metaData());
-            // check and write changes in indices
-            for (IndexMetaWriteInfo indexMetaWrite : writeInfo) {
-                try {
-                    metaStateService.writeIndex(indexMetaWrite.reason, indexMetaWrite.newMetaData);
-                } catch (Exception e) {
-                    success = false;
-                }
+        long writeManifest(String reason, Manifest manifest) throws WriteStateException {
+            assert finished == false : TRANSACTION_COMPLETED_MSG;
+            try {
+                long generation = metaStateService.writeManifest(reason, manifest);
+                commitCleanupActions.add(() -> metaStateService.cleanupManifest(generation));
+                commitCleanupActions.forEach(Runnable::run);
```

Comment:
This relies on the fact that `commitCleanupActions` does not throw an exception, otherwise we would be rolling back after commit, which would possibly be disastrous. I wonder we should guard against that by wrapping the rollback only around the writeManifest call.

Reply:
This catches `WriteStateException` and cleanup actions could not possibly throw `WriteStateException` because it's checked and action is Runnable, which run method is not allowed to throw any exception.